F# Style Guide

cartermp opened the pull request:
Addresses #2049

cc @dsyme and @eiriktsarpalis

Eirik, I haven't incorporated the stuff we worked on yet. Thus far, the work has just been in reigning in various bits of information and trying to make it all sound good.

cartermp committed:
first cut

cartermp committed:
Finish up component guidelines

cartermp committed:
clean up markdown

cartermp committed:
formatting

cartermp committed:
start conventions

cartermp committed:
more conventions

dsyme commented:
@cartermp Why not make the updates on fsharp.org, and if necessary mirror here?

My gut feeling is that the component design guidelines should be maintained by a community, neutral organisation on a non-Microsoft site.  It feels important for the community to be owning this, and being seen to own this.

cartermp commented:
@dsyme Reasons are n-fold, in order of importance:

1. Lots of feedback from beginners that we need a style/formatting and "how do I do stuff correctly" guide. This is also echoed by feedback from @eiriktsarpalis and others in larger organizations. Such a guide does not exist on fsharp.org. Component Design Guidelines are a part of that, but they are scoped towards a specific thing.

2. Our site holds more weight and the content is more discoverable. For newcomers especially, having something up to date from a Microsoft domain will be important. Discoverability _might_ be solved by the eventual fsharp.org redesign, but I don't want to be sitting on content that I can distribute more effectively in the interim.

3. The Component Design Guidelines aren't really being maintained. [Last commit is in 2016](https://github.com/fsharp/fsfoundation/commits/gh-pages/specs/component-design-guidelines/index.md) with history indicating that the major work came from a Microsoft employee and not a community effort.

4. I'm not convinced that the visibility of this - guidelines on how to write F# code - is inherently a community thing nor something that the community particularly cares to own. I do legitimately think that the optics of Microsoft also having guidelines for how to write F# code effectively far outweighs any possible negatives (which are likely to be associated with an older Microsoft from a different time).

5. Since our documentation is open source, I think it's perfectly fine to see this as a collaboration between Microsoft and the community. After all, it is how this effort started.

That said, if there is a plan to keep the CDG on fsharp.org after the redesign and cleanup of the site, I see no reason why this couldn't also be mirrored there.

cartermp committed:
More updates to conventions

eiriktsarpalis commented on docs/fsharp/style-guide/conventions.md:
````diff
@@ -111,10 +113,148 @@ This enables the following:
+    | UndisclosedFailure -> printfn "Failed: unknown - retry perhaps?"
+```
+
+In general, if you can model the different ways that something can **fail** in your domain, then error handling code is no longer treated as something you must deal with in addition to regular program flow. It is simply a part of normal program flow. This also makes it easier to maintain as your domain changes over time.
````
might be worth mentioning that explicit error cases are more amenable to unit testing.

eiriktsarpalis commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -111,10 +113,148 @@ This enables the following:
+
+## Do not use partial application and currying in public APIs
+
+With little exception, the use of partial application in public APIs can be confusing for consumers. Generally speaking, `let`-bound values in F# code are **values**, not **function values**. Mixing together values and function values can result in saving a very small number of lines of code in exchange for quite a bit of cognitive overhead, especially if combined with operators such as `>>` to c ...
```
A point to drive through here is that curried functions do not label their arguments, so instead of
```fsharp
val foo : name:string -> age:int -> unit
```
you get
```fsharp
val foo : (string -> int -> unit)
```
which can be confusing from a documentation perspective.

eiriktsarpalis replied:
Recommended workaround would be to do full-blown eta-expansion, even if it does feel redudant.

cartermp replied:
Covered.

cartermp committed:
Update partial application section and respond to feedback

cartermp committed:
add tooling and point-free, access control, type inference, generics

cartermp committed:
perf section

cartermp committed:
Fix warnings

eiriktsarpalis commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,455 @@
+    let dep3 = Random().Next()
+ 
+    let function1 arg = doStuffWith dep1 dep2 dep3 arg
+    let function2 arg = doSutffWith dep1 dep2 dep3 arg
```
Few other notes around module design:
* Public module functions should rarely accept more than 3-4 arguments.
* Public module functions should not accept too many function arguments. If they do expose function arguments, their intent should be evident simply by examining the function signature (think hoogle). I cannot understand what this function does : `val myCoolFunc : (int -> bool -> string -> int) -> (int -> 'T -> string) -> string * 'T list`.
* Mostly avoid exposing third-order functions or higher, e.g. `val foo : ((string -> unit) -> string) -> string`.
* If your function still requires too many function arguments, consider refactoring so that it accepts an interface. This usually records intention better.
* If your functions take more than 4-5 arguments, consider converting your module into a static facade, exposing methods that name their individual parameters. Do make use of optional parameters when appropriate.
* Avoid mixing tupled arguments with curried style, e.g. `val foo : (int * string) -> int -> (bool * bool * bool) -> int`.
* Avoid using tuples as return types. Pairs and triples are occasionally acceptable, when the API clearly communicates the purpose of each element type. In most cases however, it might make sense to write a custom record type (or use anonymous records in future F# releases) that clearly labels each of the values.

bartelink commented on docs/fsharp/style-guide/component-design-guidelines.md:
````diff
@@ -0,0 +1,858 @@
+Instead of F# lists:
+
+```fsharp
+member this.PrintNames(names : list<string>) =
````
use string list to emphasize ?

bartelink commented on docs/fsharp/style-guide/component-design-guidelines.md:
````diff
@@ -0,0 +1,858 @@
+member this.TupledArguments(str, num) = String.replicate num str
+```
+
+Tip: If you’re designing libraries for use from any .NET language, then there’s no substitute for actually doing some experimental C# and Visual Basic programming to ensure that your libraries look good from these languages. You can also use tools such as .NET Reflector and the Visual Studio Object Browser to ensure that libraries and their documentation appear as expected to developers.
````
look good -> feel right/work well?

bartelink commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,858 @@
+
+* Functions and lists become references to corresponding types in the F# library.
+
+The full rules for how F# types, modules, and members are represented in the .NET Common Intermediary Language are explained in the F# language reference on the F# website.
```
Intermediate

bartelink commented on docs/fsharp/style-guide/component-design-guidelines.md:
````diff
@@ -0,0 +1,858 @@
+The code below shows how to adjust this code to take these things into account.
+
+```fsharp
+namespace ExpertFSharp.Types
````
refrain from egregious product placement please ;)

bartelink commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,858 @@
+
+* We adjusted several names: `Point1`, `n`, `l`, and `f` became `RadialPoint`, `count`, `factor`, and `transform`, respectively.
+
+* We used a return type of `seq<RadialPoint>` instead of `RadialPoint list` by changing a list construction using `[ ... ]` to a sequence construction using `seq { ... }`.
```
seq<RadialPoint> -> IEnumerable<RadialPoint> ?

bartelink reviewed (commented):
Lots of style/clarity suggestions; I've erred on the side of stating every thought the text provokes - please dont feel the need to ack or apply any of my comments

cartermp committed:
Feedback from Eirik and Ruben

cartermp commented:
Thanks for the review so far, @eiriktsarpalis and @bartelink - I applied feedback to most of your comments. Eirik, I've tracked some of your additional suggestions [in the tracking gist](https://gist.github.com/cartermp/8ce2e115f32038cf49151ff783807cc4) (I know, super high quality tool).

eiriktsarpalis commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,858 @@
+
+#### Consider using `[<RequireQualifiedAccess>]` and carefully apply `[<AutoOpen>]` attributes
+
+Adding the `[<RequireQualifiedAccess>]` attribute to a module indicates that the module may not be opened and that references to the elements of the module require explicit qualified access. For example, the Microsoft.FSharp.Collections.List module has this attribute.
```
This can be abbreviated to `FSharp.Collections.List`

eiriktsarpalis commented on docs/fsharp/style-guide/conventions.md:
````diff
@@ -191,9 +193,7 @@ Reconciling functionality to perform in the face of an exception with pattern ma
-It is seen as somewhat taboo in functional programming to use exceptions. Indeed, exceptions violate purity and totality, so it's safe to consider them not-quite functional. However, this ignores the reality of where code must run. Even in purely functional languages like Haskell, programs such as ```2 `div` 0``` produce a runtime exception.
-
-F# runs on .NET, and exceptions are a key piece of .NET. This is not something to ignore or attempt to work around.
+It is seen as somewhat taboo in functional programming to use exceptions. Indeed, exceptions violate purity and totality, so it's safe to consider them not-quite functional. However, this ignores the reality of where code must run. F# runs on .NET, and exceptions are a key piece of .NET. This is not something to ignore or attempt to work around.
````
Depending on who reads this, the tone of the paragraph might be interpreted as "writing pure and total function is an admirable goal, however the .NET mothership is here to crush your dreams". Would rephrase into something along the lines of "write code on the assumption that most things are neither pure or total, to minimize unpleasant surprises"

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,475 @@
+
+### Prefer `let mutable` to reference cells
+
+Reference cells can be a powerful mechanism for pointer arithemetic, but are generally not ideal for other performance-critical code. Consider the following function:
```
arithemetic. Also I have no idea why they can be "can be a powerful mechanism for pointer arithemetic" ?

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,475 @@
+let kernels =
+    let acc = ref Set.empty
+
+    ProcessWorkList startKernels (fun kernel ->
```
should `ProcessWorkList` be camel cased ? (not sure what's being conveyed)

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,475 @@
+    let mutable acc = Set.empty
+
+    ProcessWorkList startKernels (fun kernel ->
+        if not (acc.Contains(kernel) then
```
missing right paren nd/or remove the one after `Contains`

bartelink commented on docs/fsharp/style-guide/conventions.md:
````diff
@@ -0,0 +1,475 @@
+    acc |> Seq.toList
+```
+
+Aside from the single point of mutation in the middle of the lambda expression, all other code which works with `acc` no differently than if it were a normal `let`-bound immutable value. This will make it easier to change over time.
````
which -> that

>all other code which works with  `acc` no differently than if it were a normal

all other code that touches `acc` can do so in a manner that is no different to the usage of a normal

bartelink commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,506 @@
+    |> Array.concat
+
+// OK
+let methods2 = System.AppDomain.CurrentDomain.GetAssemblies()
```
now this deserves Tolerable ;) as it invariably degrades to the next form

bartelink commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+
+```fsharp
+acc +
+(Printf.sprintf "\t%s - %i\n\r"
````
Does the `Printf.` serve a purpose

bartelink commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,506 @@
+    List.iter (fun elem ->
+        printfn "%d" (a + elem)) list1
+
+// Tolerable
```
tolerable is not a pejorative used elsewhere within this doc

bartelink commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,506 @@
+
+In general, most function parameter application is done on the same line.
+
+If you wish to apply parameters to a function on a new line, indent them 4 lines.
```
indent by one step

bartelink commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,506 @@
+
+### Formatting modules
+
+Code in a local module must be indented relative to the module, but code in a top-level module does not have to be indented. Namespace elements do not have to be indented.
```
does not have to be -> should not be

bartelink commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+```fsharp
+let comparer =
+    { new IComparer<string> with
+          member x.Compare(s1, s2) =
````
this is indented by 5

cartermp replied:
😱

bartelink commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,506 @@
+          member x.Compare(s1, s2) =
+              let rev (s : String) =
+                  new String (Array.rev (s.ToCharArray()))
+              let reversed = rev s1 i
```
i ?

bartelink commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+spam ( ham.[ 1 ] )
+```
+
+Named arguments should also not have space surrounding the `=`:
````
I find this to be contentious though I don't dislike it. Might be worth putting two args into the example in order to illustrate that commas as separators are delimiters and hence would have a space after them

bartelink commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+*)
+```
+
+Start inline comments with a new line, and captialize the first letter.
````
captialize

bartelink commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+Start inline comments with a new line, and captialize the first letter.
+
+```fsharp
+let f x = x + 1 // Increment by one.
````
'start inline comments with a new line" ? Can we indicate if this is bad (and if so, what's the ideal) ?
IMO sometimes an EOL // comment is appropriate and an important counterpoint to comments level with the code, which I'd normally use to indicate aht's about to happen

bartelink commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+    ...
+```
+
+In large assemblies and implementation files, PascalCase is sometimes used to indicate major internal routines.
````
I don't think there's a need to promulgate this as the value of such a scheme is minimal, versus the risk of people using this to further arbitrary inconsistency. There are many better ways to achieve similar demarcation. Next we'll be advocating #regions !

cartermp replied:
Deaded it

bartelink commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,37 @@
+
+1. **Good F# code is succinct and expressive**
+
+    F# has many features which allow you to express actions in fewer lines of code and re-use generic functionality. The F# core library also contains many useful types and functions for working with common collections of data. As a general rule, if you can express a solution to a problem in less lines of code, other developers (or your future self) will be appreciative. It is also highly recomme ...
```
which -> that

bartelink commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,37 @@
+
+3. **Good F# code makes use of object programming, not object orientation**
+
+    F# has full support for programming with objects in .NET, including [classes](../language-reference/classes.md), [interfaces](../language-reference/interfaces.md), [access modifiers](../language-reference/access-control.md), [abstract classes](../language-reference/abstract-classes.md), and so on. For more complicated functional code, such as functions which must be context-aware, objects can ...
```
which -> that
consuption

bartelink commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,37 @@
+
+5. **Good F# code is toolable**
+
+    Tools are invaluable for working in large codebases, and there are ways to write F# code such that it can be used more effectively with tools. One example is making sure you don't overdo it with a point-free style of programming, so that intermmediate values can be inspected with a debugger. Another example is using [XML documentation comments](../language-reference/xml-documentation.md) desc ...
```
intermmediate -> intermediate

bartelink commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,37 @@
+
+The [F# Formatting Guidelines](formatting.md) provide guidance on how to format code so that it is easy to read.
+
+The [F# coding conventions](conventions.md) provide guidance for F# programming idioms which will help the long-term maintenance of larger F# codebases.
```
which -> that

cartermp commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,475 @@
+3. Errors in initialization for dependent values are not likely to manifest as a `TypeInitializationException`.
+4. The API is now easier to test.
+
+One could even argue that the example with classes is more "functional" than the example with let-bound values in a module.
```
Just going to remove this line

cartermp commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,475 @@
+
+First, it is recommended that you read the [Exception Design Guidelines](../../standard/design-guidelines/exceptions.md). They are also applicable to F#.
+
+There are four primary ways to raise exceptions in F#:
```
Three!

cartermp commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,506 @@
+let point = { X = 1.0; Y = 0.0 }
+```
+
+Records which are longer should use new lines for labels:
````
I'm horrible, I know

cartermp commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,506 @@
+
+A tuple is parenthesized and the commas therein (delimiters) are each followed by a space e.g. `(1, 2)`, `(x, y, z)`.
+
+A commonly accepted exception is to omit parentheses in pattern matching of tuples. The justification is to match multiple values, not construct new tuples.
```
I'll just omit that part and show rather than tell

cartermp committed:
Feedback from Eirik and Ruben

cartermp committed:
Spelling

eiriktsarpalis commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -306,21 +306,21 @@ let tryReadAllTextIfPresent (path : string) =
-Instead of functioning as a catch-all, this function will now properly handle the case when a file was not found and assign that meaning to a return value which can map to that error case, while not discarding any contextual information or forcing callers to deal with a case that may not be relevant at that point in the code.
+Instead of functioning as a catch-all, this function will now properly handle the case when a file was not found and assign that meaning to a return . This return value can map to that error case, while not discarding any contextual information or forcing callers to deal with a case that may not be relevant at that point in the code.
 
 Types such as `Result<'Success, 'Error>` are appropriate for basic operations where they aren't nested, and F# optional types are perfect for representing when something could either reutrn *something* or *nothing*. They are not a replacement for exceptions, though, and should not be used in an attempt to replace exceptions. Rather, they should be applied judiciously to address specific aspects o ...
```
reutrn -> return

eiriktsarpalis commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,855 @@
+
+In addition to the previous table, be aware of the following:
+
+#### Avoid abbreviaitons
```
typo

cartermp committed:
Spelling

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,529 @@
+
+Note that a line break separates topological layers, with each layer being sorted alphanumerically afterwards. This cleanly organizes code without accidentally shadowing values.
+
+## Use classes to contain values which have side effects
```
that

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,529 @@
+| Function | Syntax | Purpose |
+|----------|--------|---------|
+| `invalidArg` | `invalidArg "argumentName" "message"` | Raises a `System.ArgumentException` with a specified argument name and message. |
+| `nullArg` | `nullArg "argumentName"` | Raises a `System.ArgumentNullException` with the specified argument name. |
```
move this up one - ANE is preferred to AE if it pertains

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,529 @@
+| `invalidArg` | `invalidArg "argumentName" "message"` | Raises a `System.ArgumentException` with a specified argument name and message. |
+| `nullArg` | `nullArg "argumentName"` | Raises a `System.ArgumentNullException` with the specified argument name. |
+| `invalidOp` | `invalidOp "message"` | Raises a `System.InvalidOperationException` with the specified message. |
+|`raise`| `raise ExceptionType("message")` | General purpose mechanism for throwing exceptions. |
```
`raise ExceptionType("message")`

actually requires parentheses I think? suggest: `raise (MyException("Message"))`

cartermp committed:
Feedback from Ruben

cartermp committed:
Some cleanup and basic objects section

bartelink commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -526,4 +531,79 @@ let kernels =
+
+[Object Expressions](../language-reference/object-expressions.md) allow you to implement interfaces on the fly, binding the implemented interface to a value without needing to do so inside of a class. This is very convenient, especially if you _only_ need to implement the interface and have no need for a full class.
+
+For example, here is the code which is run in Ionide to provide a code fix action if you've added a symbol that you don't have an `open` statement for:
```
that ;) (or remove "which is run")

cartermp committed:
update objs

cartermp committed:
Updates per feedback

dsyme commented:
@cartermp yes, it makes sense for Microsoft to host this guidance

I'll review after a bit

cartermp committed:
Updates as per acrolynx

cartermp committed:
Fix the bad

cartermp committed:
Add type abbreviations

cartermp committed:
Feedback from Ruben

cartermp committed:
More feedback

cartermp committed:
i can spell gud

wallymathieu committed:
Adding a section on static members

When you have static members, sometimes they mimic module functions. Having capitalization based on implementation details makes for naming inconsistencies when consuming API's in f#.

wallymathieu committed:
Fixes after review by @cartermp

wallymathieu committed:
escape [<CompiledName>]

Also clarified the role of CompiledName attribute. You can see an examples of the attribute usage in https://github.com/fsharp/fsharp/blob/master/src/fsharp/FSharp.Core/option.fs#L52

cartermp committed:
Merge pull request #1 from wallymathieu/patch-1

Adding a section on static members

dsyme commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -245,6 +245,24 @@ let u = v * 10.0
+    member v.Y = y
+    
+    [<CompiledName("Create")>]
+    static member create x y = Vector (x, y)
```
For a member, the recommendation is always to use a PascalCase name.  We would not recommend people write code like this.

dsyme commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,873 @@
+
+#### Use method overloading for member functions, if doing so provides a simpler API
+
+Method overloading is a powerful tool for simplifying an API that may need to do multiple related things.
```
In F#, it is more common to overload on number of arguments rather than types of arguments. We should probably mention this.

cartermp replied:
Added this after the code sample, which does overload on number of arguments so as to cement this as the way to do things (generally)

cartermp committed:
Update closure1 table

cartermp committed:
Merge branch 'style' of https://github.com/cartermp/docs into style

ForNeVeR commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,873 @@
+# F# component design guidelines
+
+This document is a set of component design guidelines for F# programming, based on [the F# Component Design Guidelines, v14, Microsoft Research, and [another version](http://fsharp.org/specs/component-design-guidelines/) originally curated and maintained by the F# Software Foundation.
```
Something's wrong here: square brackets are not balanced.

ForNeVeR commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,873 @@
+|-----------|------|------|----------|-------|
+| Concrete types | PascalCase | Noun/ adjective | List, Double, Complex | Concrete types are structs, classes, enumerations, delegates, records, and unions. Though type names are traditionally lowercase in OCaml, F# has adopted the .NET naming scheme for types.
+| DLLs           | PascalCase |                 | Fabrikam.Core.dll |  |
+| Union tags     | PascalCase | Noun | Some, Add, Success | Do not use a prefix in public APIs. Optionally use a prefix when internal, such as `type Teams = TAlpha | TBeta | TDelta.` |
```
Unfortunately, GitHub doesn't render `` `type Teams = TAlpha | TBeta | TDelta.` |`` fragment properly. Will docs.microsoft.com render it?

cartermp committed:
Feedback and try triple ticks

cartermp committed:
update section with TODO with intention of adding code sample later

cartermp commented:
@mairaw and @BillWagner would love to get a review to merge soon (ideally before build)

BillWagner commented:
I will review tonight.

mairaw reviewed (changes requested):
Thanks @cartermp. Only reviewed 1st topic so far. Looks good but left some comments to improve it.

mairaw commented on docs/fsharp/style-guide/conventions.md:
````diff
@@ -0,0 +1,637 @@
+        | (false, _) -> false
+```
+
+`Closure1Table` encapsulates the underlying mutation-based data structure, thereby not forcing callers to maintain a the underlying data structure. Classes are a powerful way to encapsulate data and routines that are mutation-based without exposing the details to callers.
````
typo: a the

mairaw commented on docs/fsharp/style-guide/conventions.md:
````diff
@@ -0,0 +1,637 @@
+        }
+```
+
+Because there is no need for a class when interacting with the Visual Studio Code API, Object Expressions are an ideal tool for this. They are also valuable for unit testing, when you want to stub out an interface with test routines in an ad-hoc manner.
````
ad-hoc -> ad hoc

mairaw commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,637 @@
+
+### Avoid using type abbreviations to represent your domain
+
+Although Type Abbreviations are convenient for giving a name to function signatures, they can be confusing when abbreviating other types. Consider this abbreviation:
```
type abbreviations capitalization is not consistent

mairaw commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,505 @@
+
+Write `x :: l` with spaces around the `::` operator (`::` is an infix operator, hence surrounded by spaces) and `[1; 2; 3]` (`;` is a delimiter, hence followed by a space).
+
+Always use at least one space between two distinct brace-like operators (e.g. leave a space between a `[` and a `{`).
```
e.g. -> for example

better to start a new sentence instead of putting the sentence in parentheses.

mairaw commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,505 @@
+
+* Separate top-level function and class definitions with two blank lines.
+* Method definitions inside a class are separated by a single blank line.
+* Extra blank lines may be used (sparingly) to separate groups of related functions. Blank lines may be omitted between a bunch of related one-liners (e.g. a set of dummy implementations).
```
e.g. -> for example

mairaw commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,505 @@
+
+### Use prefix syntax for generics (`Foo<T>`) in preference to postfix syntax (`T Foo`)
+
+F# inherits both the postfix ML style of naming generic types, e.g. `int list` as well as the prefix .NET style, e.g. `list<int>`. You should prefer the .NET style, except for four specific types. For F# lists, use the postfix form: `int list` rather than `list<int>`. For options, use the postfix form: `int option` rather than `option<int>`. For arrays, use the syntactic name `int[]` rather than  ...
```
e.g. -> for example

You should prefer seems an odd sentence. wondering if the four specific types should be on a list.

mairaw commented on docs/fsharp/style-guide/formatting.md:
````diff
@@ -0,0 +1,505 @@
+
+The following operators are defined in the F# standard library and should be used instead of defining equivalents. Using these operators is recommended as it tends to make code more readable and idiomatic. Developers with a background in OCaml or other functional programming language may be accustomed to different idioms, and this list summarizes the recommended F# operators.
+
+```fsharp
````
content inside code blocks doesn't get translated. Perhaps a table would be better?

mairaw commented on docs/fsharp/style-guide/formatting.md:
```diff
@@ -0,0 +1,505 @@
+
+### Use standard F# operators
+
+The following operators are defined in the F# standard library and should be used instead of defining equivalents. Using these operators is recommended as it tends to make code more readable and idiomatic. Developers with a background in OCaml or other functional programming language may be accustomed to different idioms, and this list summarizes the recommended F# operators.
```
this list -> the following list (or table if you change to a table) and end with :

Consider making that last part a new sentence

mairaw commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,37 @@
+
+The following articles describe guidelines for formatting F# code and topical guidance for features of the language and how they should be used.
+
+This guidance has been formulated based on the use of F# in large codebases with a diverse group of programmers. We have found that this guidance will generally lead to successful use of F# and minimize frustrations when requirements for programs change over time.
```
remove we

mairaw commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,37 @@
+
+    Tools are invaluable for working in large codebases, and there are ways to write F# code such that it can be used more effectively with tools. One example is making sure you don't overdo it with a point-free style of programming, so that intermediate values can be inspected with a debugger. Another example is using [XML documentation comments](../language-reference/xml-documentation.md) descr ...
+
+## Read next
```
I believe the new templates call this section Next steps (can't access them right now). Or we usually call this See also.

cartermp commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,875 @@
+
+#### Hide the representations of record and union types if the design of these types is likely to evolve
+
+Avoid revealing concrete representations of objects. For example, the concrete representation of **System.DateTime** values is not revealed by the external, public API of the .NET library design. At runtime the Common Language Runtime knows the committed implementation that will be used throughout execution. However, compiled code does not itself pick up dependencies on the concrete representatio ...
```
Heh, I like how this is a link to a style guide 😄

ReedCopsey commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,875 @@
+# F# component design guidelines
+
+This document is a set of component design guidelines for F# programming, based on the F# Component Design Guidelines, v14, Microsoft Research, and [another version](http://fsharp.org/specs/component-design-guidelines/) originally curated and maintained by the F# Software Foundation.
```
@cartermp https should now work, and render correctly. Github pages finally supports it, so we were in the process of switching, and had a resource bug. It's now fixed.

cartermp committed:
Maira feedback

cartermp committed:
fixes

BillWagner commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,648 @@
+
+With little exception, the use of partial application in public APIs can be confusing for consumers. Usually, `let`-bound values in F# code are **values**, not **function values**. Mixing together values and function values can result in saving a small number of lines of code in exchange for quite a bit of cognitive overhead, especially if combined with operators such as `>>` to compose functions ...
+
+### Consider the tooling implications for point-free programming
```
the term "point-free programming" is new to me. Will your target audience know it?

cartermp replied:
Yep! Well, mostly. Certainly FP programmers who come to F#.

BillWagner commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,648 @@
+
+F# values are immutable by default, which allows you to avoid certain classes of bugs (especially those involving concurrency and parallelism). However, in certain cases, in order to achieve optimal (or even reasonable) efficiency of execution time or memory allocations, a span of work may best be implemented by using in-place mutation of state. This is possible in an opt-in basis with F# with th ...
+
+However, use of `mutable` in F# may feel at odds with functional purity. This is fine, if you adjust expectations from purity to referential transparency. We argue that referential transparency - not purity - is the end goal when writing F# functions. This allows you to write a functional interface over a mutation-based implementation for performance critical code.
```
Should you define "referential transparency"?

BillWagner commented on docs/fsharp/style-guide/conventions.md:
```diff
@@ -0,0 +1,648 @@
+* Named and Optional arguments
+* Interfaces and interface implementations
+
+**Don't reach for these features first, but do judiciously apply them when they make sense:**
```
I think you need to expound on "when they make sense". Without some detail on when it makes sense, this isn't very helpful guidance.

BillWagner commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,48 @@
+
+The following articles describe guidelines for formatting F# code and topical guidance for features of the language and how they should be used.
+
+This guidance has been formulated based on the use of F# in large codebases with a diverse group of programmers. This guidance will generally lead to successful use of F# and minimize frustrations when requirements for programs change over time.
```
"will generally lead" => "generally leads"

BillWagner commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,48 @@
+
+## Five principles of good F# code
+
+The following principles should be kept in mind any time you write F# code, especially in systems that will change over time. Every piece of guidance in further articles stems from these five points.
```
"You should keep the following principles in mind" would be a stronger start.

BillWagner commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,48 @@
+
+4. **Good F# code performs well without exposing mutation**
+
+    It's no secret that to write high-performance code, you must use mutation. It's how computers work, after all. Such code is often error-prone and difficult to get right. Exposing that mutation to callers should be avoided. A functional interface over a mutation-based implementation should be sought instead.
```
a couple instance of passive voice here: 

Instead of "Exposing that mutation to callers should be avoided." try "Avoid exposing mutation to callers".

Instead of "A functional interface over a mutation-based implementation should be sought instead." try "Seek a functional interface over a mutation-based implementation."

BillWagner commented on docs/fsharp/style-guide/index.md:
```diff
@@ -0,0 +1,48 @@
+
+5. **Good F# code is toolable**
+
+    Tools are invaluable for working in large codebases, and there are ways to write F# code such that it can be used more effectively with tools. One example is making sure you don't overdo it with a point-free style of programming, so that intermediate values can be inspected with a debugger. Another example is using [XML documentation comments](../language-reference/xml-documentation.md) descr ...
```
"..., and the there are ways" consider: "... you can write F# code..."

BillWagner reviewed (changes requested):
This is very useful stuff @cartermp I learned quite a bit reading it. Overall, this is a great set of guidance.

I had a number of suggestions for style, and a few locations where the guidance could be more actionable.

cartermp commented on docs/fsharp/style-guide/component-design-guidelines.md:
````diff
@@ -0,0 +1,875 @@
+    member this.MyEvent = myEv.Publish
+```
+
+#### Expose asynchronous operations as methods returning .NET tasks (`Task<T>`)
````
At this point, I don't think I'll call these out specifically until we have better language-level support for them in F#.

cartermp commented on docs/fsharp/style-guide/component-design-guidelines.md:
````diff
@@ -0,0 +1,882 @@
+    else ()
+```
+
+#### Avoid using tuples as return values
````
This is for the "general .NET" section, where we want to discourage it. But for F#-only consumers, it's okay for small return types as-documented earlier in the article

cartermp commented on docs/fsharp/style-guide/component-design-guidelines.md:
```diff
@@ -0,0 +1,882 @@
+
+* Functions and lists become references to corresponding types in the F# library.
+
+The full rules for how F# types, modules, and members are represented in the .NET Common Intermediate Language are explained in the F# language reference on the F# website.
```
Going to just axe this sentence. The language reference is in this doc set.

cartermp committed:
Feedback from Bill

cartermp committed:
Fix warnings

cartermp committed:
Fix link issue

selketjah commented:
I went through the style guide today. It is really great, there is a lot to like in it.
I have the following suggestions:

**Language**
Not everybody is native English, you want want to go over the text and see if some words can be replaced with more commonly known definitions.

For example:
    * succinct (alternatives: concise, least amount of words possible)
    * legible (alternatives: readable, easy to read)

**Type annotations**
different uses of spaces in the type annotations: all possibilities are present in the examples right now
let doStuff (x:int)
let doStuff (x : int)
let doStuff (x: int)

I don't know if it is being left to personal preference, but it might be a good idea to make a recommendation (or make it consistent in the style guide).
